File: example/example_sim3_solver.cc
#include <iostream>
#include <random>
#include <thread>

// GSRAP
#include "gsrap/macros.h"
#include "gsrap/sim3_solver.h"

// common
#include "common/utils.h"

// NOLINTNEXTLINE
using namespace gsrap;

// NOLINTNEXTLINE
using namespace gsrap::example;

namespace gsrap::example {

using Pair     = std::pair<uint32_t, uint32_t>;
using Iterator = std::vector<Pair>::const_iterator;

static void RunSim3Solver(void) {
  const uint32_t num_points = 500;  // Number of points

  const Vec3d center(0.0, 0.0, 0.0);  // Center of spherical point cloud.
  const double edge_length = 1;       // Edge length of cuboid point cloud

  const double gt_scale = 1.25;
  Mat3d gt_r =
      Eigen::AngleAxis<double>(
          0.25 * kPi,
          (Vec3d::UnitX() + Vec3d::UnitY() + Vec3d::UnitZ()).normalized())
          .toRotationMatrix();
  const Vec3d gt_t(1.0, 0.25, 0.5);

  std::mt19937 engine(123456789);
  eigen_vector<Vec3d> points1 = CreateCuboidlPointCloud(
      center, edge_length, edge_length, edge_length, num_points, &engine);

  eigen_vector<Vec3d> points2;
  points2.resize(num_points);
  std::transform(points1.begin(), points1.end(), points2.begin(),
                 [&gt_scale, &gt_r, &gt_t](const Vec3d& p1) {
                   return gt_scale * gt_r * p1 + gt_t;
                 });

  const std::vector<Pair> matches = CreateMatches(num_points, 1.0, &engine);

  Sim3SolverPolicy sim3_solver_policy = {};

  sim3_solver_policy.flags =
      SIM3_SOLVER_POLICY_FLAGS_NONE | SIM3_SOLVER_POLICY_FLAGS_REFINE;
  sim3_solver_policy.ransac_policy = {};
  sim3_solver_policy.ransac_policy.flags =
      RANSAC_POLICY_FLAGS_EARLY_STOP |
      RANSAC_POLICY_FLAGS_USE_PROBABILITY_WITHOUT_DUPLICATION_SAMPLE;
  sim3_solver_policy.ransac_policy.probability = 0.999999;
  sim3_solver_policy.ransac_policy.num_threads =
      std::thread::hardware_concurrency();
  sim3_solver_policy.ransac_policy.num_ransac_itr_lower_limit = 10;
  sim3_solver_policy.ransac_policy.num_ransac_itr_upper_limit = 1000000;

  sim3_solver_policy.inlier_thr = 1e-12;

  const std::pair<std::optional<Sim3SolverResult<Iterator>>, RansacReport>
      result = ComputeSim3Transformation(sim3_solver_policy, points1, points2,
                                         matches.begin(), matches.end());

  if (result.first) {
    std::cout << "scale:\n" << result.first->scale << "\n" << std::endl;
    std::cout << "gt scale:\n" << gt_scale << "\n" << std::endl;

    std::cout << "R :\n" << result.first->rotation << "\n" << std::endl;
    std::cout << "gt R :\n" << gt_r << "\n" << std::endl;

    std::cout << "t :\n" << result.first->translation << "\n" << std::endl;
    std::cout << "gt t :\n" << gt_t << "\n" << std::endl;

    std::cout << "inlier rate: " << result.first->inlier_ratio << "\n"
              << std::endl;
  }

  std::cout << "RANSAC itrations: " << result.second.num_iteration << "\n"
            << std::endl;
}

}  // namespace gsrap::example

int main(void) {
  RunSim3Solver();
  return 0;
}

File: example_py/example_sim3_solver.py
import multiprocessing
from typing import cast

import gsrap
import numpy as np
from common.utils import create_cuboid_point_cloud
from scipy.spatial.transform import Rotation


def main():
    num_points = 5000

    center = np.array([0, 0, 0])  # Center of spherical point cloud.
    edge_length = 1.0  # Edge length of cuboid point cloud

    gt_scale = 1.25

    axis = np.array([1, 1, 1], dtype=np.double)
    axis /= np.linalg.norm(axis, ord=2)
    angle = 0.25 * np.pi

    gt_r = Rotation.from_rotvec(angle * axis).as_matrix()

    gt_t = np.array([1.0, 0.25, 0.5])

    points1 = create_cuboid_point_cloud(
        center, edge_length, edge_length, edge_length, num_points, 123456
    )

    points2 = (gt_scale * gt_r @ points1.transpose()).transpose() + gt_t

    matches = np.arange(num_points, dtype=np.uint32)[:, None].repeat(2, axis=1)

    sim3_solver_policy = gsrap.Sim3SolverPolicy()

    sim3_solver_policy.flags = int(
        gsrap.Sim3SolverPolicyFlags.SIM3_SOLVER_POLICY_FLAGS_NONE
    ) | int(gsrap.Sim3SolverPolicyFlags.SIM3_SOLVER_POLICY_FLAGS_REFINE)

    sim3_solver_policy.ransac_policy.flags = int(
        gsrap.RansacPolicyFlags.RANSAC_POLICY_FLAGS_EARLY_STOP
    ) | int(
        gsrap.RansacPolicyFlags.RANSAC_POLICY_FLAGS_USE_PROBABILITY_WITHOUT_DUPLICATION_SAMPLE
    )
    sim3_solver_policy.ransac_policy.num_threads = multiprocessing.cpu_count()
    sim3_solver_policy.ransac_policy.num_ransac_itr_lower_limit = 10
    sim3_solver_policy.ransac_policy.num_ransac_itr_upper_limit = 1000000

    sim3_solver_policy.inlier_thr = 1e-12
    gsrap.compute_sim3_transformation(sim3_solver_policy, points1, points2, matches)
    ret = gsrap.compute_sim3_transformation(
        sim3_solver_policy, points1, points2, matches
    )

    sim3_solver_result = ret[0]
    if sim3_solver_result is not None:
        sim3_solver_result = cast(gsrap.Sim3SolverResult, sim3_solver_result)
        print("scale:\n", sim3_solver_result.scale, "\n")
        print("gt scale:\n", gt_scale, "\n")

        print("R :", sim3_solver_result.rotation, "\n")
        print("gt R :", gt_r, "\n")

        print("t :", sim3_solver_result.translation, "\n")
        print("gt t :", gt_t, "\n")

    print("RANSAC interation: ", ret[1].num_iteration)


if __name__ == "__main__":
    main()
